### PR TITLE
Add GCP token validation

### PR DESCRIPTION
**Pull request by atburke**
This PR adds validation for ID tokens generated for GCP VMs. Part of #22815.

Should merge after #23946 and before #24493.

### CODE REVIEW

File: lib/gcp/gcp.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	// The instance's zone.
+	Zone string `json:"zone"`
+	// The instance's ID.
+	InstanceID string `json:"instance_id"`
+	// The instance's name.
+	InstanceName string `json:"instance_name"`
+}
+
+// Google contains Google-specific token claims.
+type Google struct {
```
Comment (zmb3) [inline lib/gcp/gcp.go:35]: Do these struct types need to be exported?

File: lib/gcp/gcp.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	InstanceName string `json:"instance_name"`
+}
+
+// Google contains Google-specific token claims.
+type google struct {
+	ComputeEngine computeEngine `json:"compute_engine"`
+}
+
+// IDTokenClaims is the set of claims in a GCP ID token.
+type IDTokenClaims struct {
```
Comment (strideynet) [inline lib/gcp/gcp.go:40]: Please provide a link to the documentation for this ID token's claims in the godoc, it makes double checking these values are correct much easier.

File: lib/gcp/gcp.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+// Google contains Google-specific token claims.
+type google struct {
+	ComputeEngine computeEngine `json:"compute_engine"`
+}
+
+// IDTokenClaims is the set of claims in a GCP ID token. GCP documentation for
+// claims can be found at
+// https://cloud.google.com/compute/docs/instances/verifying-instance-identity#payload
+type IDTokenClaims struct {
```
Comment (strideynet) [inline lib/gcp/gcp.go:42]: Can we follow the example of `circleci`, `github` and `gitlab` and implement `JoinAuditAttributes()` for this type so we can provide better join audit information about hosts that have joined ?

**Review by strideynet — APPROVED** [review]

File: lib/gcp/token_validator.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/jwt"
+)
+
+type IDTokenValidatorConfig struct {
```
Comment (mdwn) [inline lib/gcp/token_validator.go:31]: nit: godoc.

File: lib/gcp/token_validator.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	Clock clockwork.Clock
+	// issuerHost is the host of the Issuer for tokens issued by Google, to be
+	// overridden in tests. Defaults to "accounts.google.com".
+	issuerHost string
+	// insecure configures the validator to use HTTP rather than HTTPS, to be
+	// overridden in tests.
+	insecure bool
+}
+
+type IDTokenValidator struct {
```
Comment (mdwn) [inline lib/gcp/token_validator.go:43]: nit: godoc.

File: lib/gcp/token_validator.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+func (id *IDTokenValidator) issuerURL() string {
+	scheme := "https"
+	if id.insecure {
+		scheme = "http"
+	}
+	return fmt.Sprintf("%s://%s", scheme, id.issuerHost)
+}
+
+func (id *IDTokenValidator) Validate(ctx context.Context, token string) (*IDTokenClaims, error) {
```
Comment (mdwn) [inline lib/gcp/token_validator.go:67]: nit: godoc.

**Review by mdwn — APPROVED** [review]
Just a few nits, but LGTM. Approving!

**Pull request merged**